packages/react/use-body-pointer-events/src/useBodyPointerEvents.tsx

```diff
@@ -1,21 +1,76 @@
+  const isMousePressedRef = React.useRef(false);
+
+  React.useEffect(() => {
+    const handlePointerDown = (event: PointerEvent) => {
+      const isMouse = event.pointerType === 'mouse';
+      isTouchOrPenPressedRef.current = !isMouse;
+      isMousePressedRef.current = isMouse;
+    };
+    const handlePointerUp = () => {
+      isTouchOrPenPressedRef.current = false;
+      isMousePressedRef.current = false;
+    };
+    document.addEventListener('pointerdown', handlePointerDown);
+    document.addEventListener('pointerup', handlePointerUp);
+    return () => {
+      document.removeEventListener('pointerdown', handlePointerDown);
+      document.removeEventListener('pointerup', handlePointerUp);
+    };
+  }, [isTouchOrPenPressedRef]);
+
   useLayoutEffect(() => {
     if (disabled) {
       if (changeCount === 0) {
         originalBodyPointerEvents = document.body.style.pointerEvents;
       }
+
+      function resetPointerEvents() {
+        if (changeCount === 0) {
+          document.body.style.pointerEvents = originalBodyPointerEvents;
+        }
+      }
+
       document.body.style.pointerEvents = 'none';
       changeCount++;
 
       return () => {
         changeCount--;
-        if (changeCount === 0) {
-          document.body.style.pointerEvents = originalBodyPointerEvents;
+        if (isTouchOrPenPressedRef.current) {
+          /**
+           * We force pointer-events to remain disabled until `click` fires on touch devices
+           * because browsers implement a ~350ms delay between the time the user stops
+           * touching the display and when the browser executes events. We need to ensure we
+           * don't reactivate pointer-events within this timeframe otherwise the browser may
+           * execute events that should have been prevented.
+           *
+           * We are aware that `touch-action: manipulation` shortens this delay for events,
+           * but it isn't enough to cover all cases.
+           *
+           * When there is an input on screen:
+           * - if a click event is bound to it, it will fire after a `pointerdown` which may
+           * have re-enabled pointer-events (regardless of `touch-action: manipulation`).
+           * - if clicking it causes the page to zoom, the events will wait for the zoom to
+           * finish before executing on the input.
+           * - if long pressesing it, the events will execute after the longpress delay.
+           */
+          document.addEventListener('click', resetPointerEvents, { once: true });
+        } else if (isMousePressedRef.current) {
+          /**
+           * We force pointer-events to remain disabled until `pointerup` in-case a mouse pointer
+           * down re-enables pointer-events. Otherwise, events bound to inert controls could
+           * execute after pointer-events have been re-enabled, e.g. `select` event.
+           */
+          document.addEventListener('pointerup', resetPointerEvents, { once: true });
```

Comment:
Fixes this issue https://modulz.slack.com/archives/C01T7FAM3CP/p1625499762023000?thread_ts=1625495026.018900&cid=C01T7FAM3CP